File: rebuild_msi.js
#!/usr/bin/env node
/**
 * Скрипт для пересборки MSI файла с встроенным JSON
 * Версия для Node.js (совместимость с WebContainer)
 */
const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');

function rebuildMsi() {
    console.log("Пересборка MSI файла...");
    
    // Проверяем наличие необходимых файлов
    const requiredFiles = [
        'Binary.idt',
        'CustomAction.idt', 
        'InstallExecuteSequence.idt',
        'Property.idt',
        'Binary/ConfigurationJson.ibd'
    ];
    
    for (const file of requiredFiles) {
        if (!fs.existsSync(file)) {
            console.log(`ОШИБКА: Файл ${file} не найден!`);
            return false;
        }
    }
    
    // Создаем резервную копию оригинального MSI
    if (fs.existsSync('install.msi')) {
        fs.copyFileSync('install.msi', 'install_backup.msi');
        console.log("Создана резервная копия: install_backup.msi");
    }
    
    try {
        // Используем msidb для обновления таблиц в MSI
        const tablesToUpdate = [
            'Binary',
            'CustomAction', 
            'InstallExecuteSequence',
            'Property'
        ];
        
        for (const table of tablesToUpdate) {
            try {
                const cmd = `msidb.exe -d install.msi -i ${table}.idt`;
                execSync(cmd, { stdio: 'pipe' });
                console.log(`Таблица ${table} успешно обновлена`);
            } catch (error) {
                console.log(`ОШИБКА при обновлении таблицы ${table}: ${error.message}`);
                return false;
            }
        }
        
        console.log("MSI файл успешно пересобран!");
        console.log("\nТеперь install.msi содержит встроенный configuration.json");
        console.log("Файл готов для развертывания через GPO или Intune");
        
        return true;
        
    } catch (error) {
        console.log(`ОШИБКА при пересборке MSI: ${error.message}`);
        return false;
    }
}

function verifyMsi() {
    console.log("Проверка содержимого MSI файла...");
    try {
        // Извлекаем список таблиц для проверки
        const listCmd = 'msidb.exe -d install.msi -l';
        const listResult = execSync(listCmd, { encoding: 'utf8' });
        
        if (listResult.includes('Binary')) {
            console.log("✓ Таблица Binary найдена в MSI");
        }
        
        // Проверяем наличие ConfigurationJson в Binary таблице
        try {
            const exportCmd = 'msidb.exe -d install.msi -e Binary';
            execSync(exportCmd, { stdio: 'pipe' });
            
            if (fs.existsSync('Binary.idt')) {
                const content = fs.readFileSync('Binary.idt', 'utf-8');
                if (content.includes('ConfigurationJson')) {
                    console.log("✓ ConfigurationJson найден в таблице Binary");
                    return true;
                }
            }
        } catch (error) {
            console.log(`Предупреждение при проверке Binary таблицы: ${error.message}`);
        }
        
        return false;
        
    } catch (error) {
        console.log(`ОШИБКА при проверке MSI: ${error.message}`);
        return false;
    }
}

// Основная функция
function main() {
    console.log("=== Пересборка MSI с встроенным JSON ===");
    
    if (rebuildMsi()) {
        console.log("\n=== Проверка результата ===");
        if (verifyMsi()) {
            console.log("\n✓ MSI файл успешно создан и проверен!");
            console.log("\nИнструкции по использованию:");
            console.log("1. Для GPO: используйте install.msi как обычно");
            console.log("2. Для Intune: загрузите install.msi в приложение Win32");
            console.log("3. JSON будет автоматически извлечен при установке");
        } else {
            console.log("\n⚠ Проверка не прошла, возможны проблемы");
        }
    } else {
        console.log("\n❌ Пересборка не удалась");
    }
}

// Запуск скрипта
if (require.main === module) {
    main();
}

module.exports = { rebuildMsi, verifyMsi };

File: rebuild_msi.py
#!/usr/bin/env python3
"""
Скрипт для пересборки MSI файла с встроенным JSON
"""
import os
import subprocess
import shutil

def rebuild_msi():
    """Пересобирает MSI файл с обновленными таблицами"""
    
    print("Пересборка MSI файла...")
    
    # Проверяем наличие необходимых файлов
    required_files = [
        'Binary.idt',
        'CustomAction.idt', 
        'InstallExecuteSequence.idt',
        'Property.idt',
        'Binary/ConfigurationJson.ibd'
    ]
    
    for file in required_files:
        if not os.path.exists(file):
            print(f"ОШИБКА: Файл {file} не найден!")
            return False
    
    # Создаем резервную копию оригинального MSI
    if os.path.exists('install.msi'):
        shutil.copy2('install.msi', 'install_backup.msi')
        print("Создана резервная копия: install_backup.msi")
    
    try:
        # Используем msidb для обновления таблиц в MSI
        tables_to_update = [
            'Binary',
            'CustomAction', 
            'InstallExecuteSequence',
            'Property'
        ]
        
        for table in tables_to_update:
            cmd = [
                'msidb.exe',
                '-d', 'install.msi',
                '-i', f'{table}.idt'
            ]
            
            result = subprocess.run(cmd, capture_output=True, text=True)
            if result.returncode != 0:
                print(f"ОШИБКА при обновлении таблицы {table}: {result.stderr}")
                return False
            else:
                print(f"Таблица {table} успешно обновлена")
        
        print("MSI файл успешно пересобран!")
        print("\nТеперь install.msi содержит встроенный configuration.json")
        print("Файл готов для развертывания через GPO или Intune")
        
        return True
        
    except Exception as e:
        print(f"ОШИБКА при пересборке MSI: {e}")
        return False

def verify_msi():
    """Проверяет содержимое MSI файла"""
    try:
        # Извлекаем список таблиц для проверки
        cmd = ['msidb.exe', '-d', 'install.msi', '-l']
        result = subprocess.run(cmd, capture_output=True, text=True)
        
        if 'Binary' in result.stdout:
            print("✓ Таблица Binary найдена в MSI")
        
        # Проверяем наличие ConfigurationJson в Binary таблице
        cmd = ['msidb.exe', '-d', 'install.msi', '-e', 'Binary']
        result = subprocess.run(cmd, capture_output=True, text=True)
        
        if result.returncode == 0:
            with open('Binary.idt', 'r', encoding='utf-8') as f:
                content = f.read()
                if 'ConfigurationJson' in content:
                    print("✓ ConfigurationJson найден в таблице Binary")
                    return True
        
        return False
        
    except Exception as e:
        print(f"ОШИБКА при проверке MSI: {e}")
        return False

if __name__ == "__main__":
    print("=== Пересборка MSI с встроенным JSON ===")
    
    if rebuild_msi():
        print("\n=== Проверка результата ===")
        if verify_msi():
            print("\n✓ MSI файл успешно создан и проверен!")
            print("\nИнструкции по использованию:")
            print("1. Для GPO: используйте install.msi как обычно")
            print("2. Для Intune: загрузите install.msi в приложение Win32")
            print("3. JSON будет автоматически извлечен при установке")
        else:
            print("\n⚠ Проверка не прошла, возможны проблемы")
    else:
        print("\n❌ Пересборка не удалась")